active/0011-jq-in-rule-engine.md

```diff
@@ -146,13 +147,67 @@ only output a single map result, and the map will then be piped to the JQ clause
 
 I am happy with this syntax now.
 
-### Introduce JQ as NIF
+### Introduce JQ as Port or NIF
 
+JQ is written in portable C as a single binary, reading command line argument
+from stdin and outputting results to the stdout. It supports Linux, OS X,
+FreeBSD, Solaris, and Windows for now, so the simplest way is to package the `jq`
+binary along with all of the emqx installation packages, and talk to `jq` using
+the [erlang port](http://erlang.org/doc/tutorial/c_port.html). For someone who
+is building emqx from source code of emqx repo, he can put the jq binary in to
+the right path according to the configuration.
 
+The second approach is NIF, with the drawback of more changes to the code (compile
+the code to a dynamic C library rather than a single binary), and safety (it
+brings down the entire erlang system on crash, and may hold up the erlang
+scheduler if it returns too late). But this way has the benefit of efficiency
```

Comment:
Long-running NIF calls can be done on a dirty scheduler, however, finding the perfect threshold of whether a call should be run on dirty/or none-dirty can be a challenge.

Reply:
It can be a hard work and I may need your help :)

I can split the large JSON to several pieces of text and call `jq:parse/2` multiple times. I will try it to see if it brings some benefits.

Reply:
> I can split the large JSON to several pieces of text and call jq:parse/2 multiple times. I will try it to see if it brings some benefits.

I think that could help in some cases but it would not be a general solution as JQ can spend an infinite amount of time on a small piece of JSON.